CI: fix NFS EBUSY error in PR test workflow

alisonshao opened the pull request:
## Summary
Fixes NFS "EBUSY: resource busy or locked" errors during `actions/checkout@v4` on self-hosted 1-gpu-runner jobs.

## Problem
Self-hosted runners on NFS-backed storage encounter checkout failures when:
- Lingering Python processes from previous test runs hold open file handles
- NFS creates `.nfs*` temporary files for open-but-deleted files
- `actions/checkout@v4` attempts to delete workspace but fails with EBUSY

Example error:
```
Error: File was unable to be removed Error: EBUSY: resource busy or locked, 
unlink '/public_sglang_ci/runner-l1c-gpu-1/_work/sglang/sglang/python/sglang/logs/.nfs2a00b5d9a59622fb00000023'
```

## Solution
Add a pre-checkout cleanup step that:
1. Kills lingering processes (python, sglang, bench_serving, pytest)
2. Force-closes any open file handles in `$GITHUB_WORKSPACE`
3. Waits 2 seconds for NFS to release locks
4. Removes all workspace files including `.nfs*` files

## Testing
This fix applies to 9 jobs using `1-gpu-runner` in `pr-test.yml`.

## Checklist
- [x] Added cleanup step before all `actions/checkout@v4` calls in 1-gpu-runner jobs
- [x] Cleanup uses `|| true` to not fail if no processes found
- [x] 2-second wait added for NFS lock release

alisonshao committed:
Fix NFS EBUSY error in PR test workflow

Add pre-checkout cleanup step to all 1-gpu-runner jobs to prevent
"EBUSY: resource busy or locked" errors caused by lingering processes
holding file handles on NFS-mounted workspaces.

The cleanup step kills lingering processes, force-closes file handles,
waits for NFS to release locks, and removes .nfs* lock files before
checkout runs.

Fixes 9 jobs that use 1-gpu-runner in pr-test.yml.

alisonshao committed:
Fix lint issues in SageMaker example file

alisonshao committed:
Add cleanup to multimodal-gen-test and 2-gpu-runner jobs

Extend NFS EBUSY fix to cover all self-hosted runner jobs:
- Added cleanup to multimodal-gen-test (1-gpu-runner)
- Added cleanup to all 2-gpu-runner jobs (3 total)

Now all 13 self-hosted runner jobs have pre-checkout cleanup.

alisonshao committed:
Remove sudo from cleanup script for runners without sudo

Some self-hosted runners don't have sudo available. Remove sudo
requirement from cleanup script since the workspace should be
owned by the runner user anyway.

alisonshao committed:
Increase wait time and improve NFS cleanup strategy

- Kill all Python processes with -9 signal for immediate termination
- Increase sleep from 2s to 5s for NFS lock release
- Explicitly remove .nfs* files before workspace cleanup
- Also remove hidden files in workspace cleanup

Kangyan-Zhou commented on .github/workflows/pr-test.yml:
```diff
@@ -208,6 +208,15 @@ jobs:
     steps:
+      - name: Kill lingering processes and cleanup workspace
+        run: |
+          pkill -f "python.*sglang" || true
```
Can we move this command to a new stage after Run Test so that this script can stored in a file in the repo instead?

Also do we need to run rm -rf "$GITHUB_WORKSPACE"/* || true? IIUC this will remove the code repo as well might slow down the code checkout time, but I'm not sure about the performance implication

Kangyan-Zhou committed:
Merge branch 'main' into fix-nfs-ebusy-error

alisonshao committed:
Use lsof with workspace root to kill file-holding processes

Key changes:
- Derive workspace root from pwd (works across all runner variants)
- Use lsof +D to find ALL processes with open files in workspace
- Kill processes gracefully first, then force kill if needed
- Wait 3s for NFS lock release
- Remove workspace contents after processes are killed

This addresses the root cause: processes holding file descriptors
in python/sglang/logs/ that create .nfs* files when deleted.

alisonshao committed:
Fix root cause: disable performance logging in CI

The NFS EBUSY error was caused by performance_logger.py creating log
files in python/sglang/logs/ during CI runs. These log files were held
open by Python processes, causing NFS to create .nfs* lock files when
checkout tried to delete the workspace.

Changes:
- Modified performance_logger.py to check SGLANG_IS_IN_CI env var
- Skip log file creation entirely when running in CI
- Added SGLANG_IS_IN_CI=true to multimodal-gen-test job
- Reverted all previous workflow cleanup attempts

This addresses the root cause rather than trying to work around it
with process cleanup scripts.

alisonshao committed:
Set SGLANG_IS_IN_CI globally for all PR test jobs

The performance_logger.py module was still creating log files in jobs
that didn't have SGLANG_IS_IN_CI set, causing NFS EBUSY errors.

Set the environment variable at the workflow level so it applies to
ALL jobs, not just multimodal-gen-test.

alisonshao committed:
Fix NFS EBUSY errors by preventing log creation in CI and adding cleanup steps

This commit implements a dual-strategy fix for NFS EBUSY errors that occur
when GitHub Actions tries to delete workspace files during checkout:

1. Prevention: Use existing is_in_ci() function from sglang.utils to disable
   performance log file creation during CI runs. This prevents the creation
   of log files that could cause NFS lock files (.nfs*).

2. Cleanup: Add post-test cleanup steps to all test jobs that run with
   if: always() to ensure logs directory is removed regardless of test
   success or failure.

alisonshao committed:
Add pre-checkout cleanup for multimodal-gen-test

Remove python/sglang/logs before checkout to prevent NFS lock files
from previous runs.

alisonshao committed:
Merge branch 'main' into fix-nfs-ebusy-error

alisonshao committed:
Kill lingering processes before cleanup in multimodal-gen-test

NFS lock files are held by lingering processes from previous runs.
Kill python/pytest processes before attempting to remove logs directory.

alisonshao committed:
Fix ordering: move Embeddings tests before cleanup step

alisonshao committed:
Comment out multimodal cleanup, remove 4/8-gpu cleanup, and disable notebook tests

1. Comment out pre-checkout and post-test cleanup for multimodal-gen-test
2. Remove cleanup steps from all 4-gpu and 8-gpu runners:
   - unit-test-backend-4-gpu
   - unit-test-backend-8-gpu-h200
   - unit-test-backend-8-gpu-h20
   - unit-test-deepep-4-gpu
   - unit-test-deepep-8-gpu
   - unit-test-backend-4-gpu-b200
   - unit-test-backend-4-gpu-gb200
3. Comment out run-all-notebooks job in execute-notebook.yml

Keep cleanup only for 1-gpu and 2-gpu runners that run performance tests.

alisonshao committed:
Revert notebook test changes - keep it active

alisonshao committed:
Merge branch 'main' into fix-nfs-ebusy-error

mickqian committed:
upd

mickqian committed:
upd

mickqian committed:
OnDemandFileHandler

mickqian committed:
clean
